feat: add C implementation for `stats/base/dists/hypergeometric/mean`

manvith2003 opened the pull request:
Resolves #3662

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for `@stats/base/dists/hypergeometric/mean` along with relevant docs, tests, examples and benchmarks.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves [[RFC]: Add C implementation for @stdlib/stats/base/dists/hypergeometric/mean #3662](https://github.com/stdlib-js/stdlib/issues/3662)

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

manvith2003 committed:
feat: add C implementation for stats/base/dists/hypergeometric/mean

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

stdlib-bot committed:
chore: update copyright years

Planeshifter commented:
This PR should be refactored to have the parameters of the distribution be typed as `int_32` instead of `double`; see other packages in https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/base/dists/hypergeometric for reference.

Planeshifter reviewed (changes requested):
This PR should be refactored to have the parameters of the distribution be typed as int_32 instead of double; see other packages in https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/base/dists/hypergeometric for reference.

Planeshifter committed:
refactor: use int32_t instead of double for hypergeometric distribution parameters

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
Merge branch 'develop' into hypergeometric_mean

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
style: add trailing newline

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
style: add trailing newline

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
style: only keep one newline

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/hypergeometric/mean/lib/native.js:
```diff
@@ -0,0 +1,80 @@
+* var v = mean( 10, 5, 12 );
+* // returns NaN
+*
+* @example
+* var v = mean( 10.3, 10, 4 );
+* // returns NaN
+*
+* @example
+* var v = mean( 10, 5.5, 4 );
+* // returns NaN
+*
+* @example
+* var v = mean( 10, 5, 4.5 );
+* // returns NaN
+*
+* @example
+* var v = mean( NaN, 10, 4 );
+* // returns NaN
+*
+* @example
+* var v = mean( 20, NaN, 4 );
+* // returns NaN
+*
+* @example
+* var v = mean( 20, 10, NaN );
+* // returns NaN
```
```suggestion
* @example
* var v = mean( 10, -2, 4 );
* // returns NaN
```

Planeshifter committed:
docs: update examples

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
docs: fix section tags

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/hypergeometric/mean/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,74 @@
+var resolve = require( 'path' ).resolve;
+var bench = require( '@stdlib/bench' );
+var Float64Array = require( '@stdlib/array/float64' );
+var randu = require( '@stdlib/random/base/randu' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var tryRequire = require( '@stdlib/utils/try-require' );
+var ceil = require( '@stdlib/math/base/special/ceil' );
+var floor = require( '@stdlib/math/base/special/floor' );
```
```suggestion
var discreteUniform = require( '@stdlib/random/base/discrete-uniform' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var tryRequire = require( '@stdlib/utils/try-require' );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/hypergeometric/mean/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,74 @@
+	K = new Float64Array( len );
+	n= new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		N[ i ] = ceil( randu() * 100.0 );
+		K[ i ] = floor( randu() * N[ i ] );
+		n[ i ] = floor( randu() * N[ i ] );
```
```suggestion
		N[ i ] = discreteUniform( 1, 100 );
		K[ i ] = discreteUniform( 1, N[ i ] );
		n[ i ] = discreteUniform( 1, N[ i ] );
```

Planeshifter committed:
bench: use discreteUniform to generate random numbers

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (approved):
Should be ready to land; thanks, @manvith2003!
